File: router/transports/wifi_transport.py
"""
WiFi Transport Layer
Handles WiFi AP setup and client connections via TCP
"""

import socket
import threading
import time
import subprocess
from typing import Dict, Optional, Any

from core.config import RouterConfig
from core.logger import get_logger
from core.transport_interface import TransportInterface
from services.routing_service import RoutingService, InterfaceType
from services.device_management_service import DeviceManagementService
from services.statistics_service import StatisticsService
from protocol import ProtocolPacket, MessageType, HandshakeManager, PacketCrypto, parse_hello_packet


class WiFiClientHandler:
    """Handles communication with a single WiFi client"""
    
    def __init__(self, client_socket: socket.socket, client_address: tuple,
                 config: RouterConfig, routing_service: RoutingService,
                 device_management: DeviceManagementService,
                 statistics_service: StatisticsService):
        self.client_socket = client_socket
        self.client_address = client_address
        self.config = config
        self.routing_service = routing_service
        self.device_management = device_management
        self.statistics_service = statistics_service
        self.logger = get_logger(f"{__name__}.{client_address[0]}")
        
        self.handshake_manager = HandshakeManager(config.router_id, config.shared_secret)
        self.device_id: Optional[str] = None
        self.assigned_address: Optional[int] = None
        self.session_crypto: Optional[PacketCrypto] = None
        self.authenticated = False
        self.running = False
        
        self.logger.debug("WiFi client handler created", address=str(client_address))
    
    def start(self):
        """Start handling client communication"""
        self.running = True
        try:
            self._handle_client()
        except Exception as e:
            self.logger.error(f"Error handling client: {e}", exc_info=True)
        finally:
            self._cleanup()
    
    def stop(self):
        """Stop handling client communication"""
        self.running = False
        try:
            self.client_socket.close()
        except:
            pass
    
    def _handle_client(self):
        """Main client handling loop"""
        self.logger.info("Started handling WiFi client")
        
        # Set socket timeout for non-blocking operations
        self.client_socket.settimeout(1.0)
        
        while self.running:
            try:
                # Receive packet
                packet = self._receive_packet()
                if packet is None:
                    continue
                
                if not self.authenticated:
                    # Still in handshake phase
                    self._handle_handshake_packet(packet)
                else:
                    # Normal communication
                    self._handle_data_packet(packet)
                    
            except socket.timeout:
                continue
            except Exception as e:
                self.logger.error(f"Error in client loop: {e}")
                break
    
    def _receive_packet(self) -> Optional[ProtocolPacket]:
        """Receive a packet from the client"""
        try:
            # Read header first
            header_data = self._receive_exact(ProtocolPacket.HEADER_SIZE)
            if not header_data:
                return None
            
            # Extract payload length from header (byte 6)
            payload_length = header_data[6]
            
            # Read payload if present
            full_data = header_data
            if payload_length > 0:
                payload_data = self._receive_exact(payload_length)
                if not payload_data:
                    return None
                full_data += payload_data
            
            # Parse complete packet
            packet = ProtocolPacket.from_bytes(full_data)
            return packet
            
        except Exception as e:
            self.logger.debug(f"Error receiving packet: {e}")
            return None
    
    def _receive_exact(self, size: int) -> Optional[bytes]:
        """Receive exact number of bytes"""
        data = b''
        while len(data) < size and self.running:
            try:
                chunk = self.client_socket.recv(size - len(data))
                if not chunk:
                    return None
                data += chunk
            except socket.timeout:
                continue
            except Exception:
                return None
        
        return data if len(data) == size else None
    
    def _handle_handshake_packet(self, packet: ProtocolPacket):
        """Handle handshake packets"""
        try:
            if packet.msg_type == MessageType.HELLO:
                # Parse HELLO packet
                router_id, device_id, nonce = parse_hello_packet(packet)
                
                if router_id != self.config.router_id:
                    self.logger.warning(f"Wrong router ID from client: {router_id}")
                    return
                
                self.device_id = device_id
                self.logger.info(f"Received HELLO from device {device_id}")
                
                # Store the client nonce
                self.handshake_manager.pending_handshakes[device_id] = {'nonce1': nonce}
                
                # Send CHALLENGE
                challenge_packet = self.handshake_manager.create_challenge_packet(device_id)
                self._send_packet(challenge_packet)
                
                self.logger.debug(f"Sent CHALLENGE to device {device_id}")
            
            elif packet.msg_type == MessageType.AUTH:
                if not self.device_id:
                    self.logger.warning("Received AUTH without HELLO")
                    return
                
                # Verify authentication
                if self.handshake_manager.verify_auth_packet(self.device_id, packet):
                    # Authentication successful, assign address
                    self.assigned_address = self.routing_service.allocate_address(self.device_id)
                    
                    # Create AUTH_ACK before completing handshake (which cleans up state)
                    auth_ack_packet = self.handshake_manager.create_auth_ack_packet(
                        self.device_id, self.assigned_address
                    )
                    
                    # Complete handshake and get session key
                    session_key = self.handshake_manager.complete_handshake(self.device_id)
                    self.session_crypto = PacketCrypto(session_key)
                    
                    # Add to routing table
                    self.routing_service.add_route(
                        device_id=self.device_id,
                        interface_type=InterfaceType.WIFI,
                        connection_handle=self.client_socket,
                        session_crypto=self.session_crypto,
                        connection_info={
                            'client_ip': self.client_address[0],
                            'client_port': self.client_address[1]
                        }
                    )
                    
                    # Send AUTH_ACK
                    self._send_packet(auth_ack_packet)
                    
                    # Mark as authenticated
                    self.authenticated = True
                    
                    # Notify device management
                    self.device_management.handle_device_connected(
                        self.device_id, 'wifi', {'address': self.client_address}
                    )
                    
                    self.device_management.handle_handshake_completed(
                        self.device_id, self.assigned_address, 'wifi'
                    )
                    
                    self.logger.info(f"Device {self.device_id} authenticated",
                                   address=f"0x{self.assigned_address:04X}")
                else:
                    # Authentication failed
                    self.device_management.handle_handshake_failed(
                        self.device_id, 'wifi', 'Authentication verification failed'
                    )
                    self.logger.warning(f"Authentication failed for device {self.device_id}")
                    self.running = False
            
        except Exception as e:
            self.logger.error(f"Error handling handshake packet: {e}", exc_info=True)
            self.statistics_service.increment_errors('wifi')
    
    def _handle_data_packet(self, packet: ProtocolPacket):
        """Handle data packets from authenticated device"""
        try:
            # Decrypt packet if needed
            if packet.flags & 0x01:  # Encrypted flag
                if self.session_crypto:
                    packet.payload = self.session_crypto.decrypt_payload(packet.payload)
            
            # Handle the message through device management
            self.device_management.handle_incoming_message(
                self.device_id, packet, 'wifi'
            )
            
        except Exception as e:
            self.logger.error(f"Error handling data packet: {e}", exc_info=True)
            self.statistics_service.increment_errors('wifi')
    
    def _send_packet(self, packet: ProtocolPacket) -> bool:
        """Send a packet to the client"""
        try:
            data = packet.to_bytes()
            self.client_socket.sendall(data)
            return True
        except Exception as e:
            self.logger.error(f"Error sending packet: {e}")
            return False
    
    def send_message(self, packet: ProtocolPacket) -> bool:
        """Send a message packet to this client"""
        try:
            # Encrypt if needed and session crypto is available
            if self.session_crypto and packet.flags & 0x01:
                packet.payload = self.session_crypto.encrypt(packet.payload)
            
            return self._send_packet(packet)
        except Exception as e:
            self.logger.error(f"Error sending message: {e}")
            return False
    
    def _cleanup(self):
        """Clean up client connection"""
        try:
            self.client_socket.close()
        except:
            pass
        
        if self.device_id:
            self.device_management.handle_device_disconnected(self.device_id)
        
        self.logger.info("WiFi client handler stopped")


class WiFiTransport(TransportInterface):
    """WiFi transport layer implementation"""
    
    def __init__(self, config: RouterConfig, routing_service: RoutingService,
                 device_management: DeviceManagementService,
                 statistics_service: StatisticsService):
        self.config = config
        self.routing_service = routing_service
        self.device_management = device_management
        self.statistics_service = statistics_service
        self.logger = get_logger(__name__)
        
        # Server state
        self.server_socket: Optional[socket.socket] = None
        self.running = False
        self.accept_thread: Optional[threading.Thread] = None
        
        # Client handlers
        self.client_handlers: Dict[str, WiFiClientHandler] = {}
        self.client_threads: Dict[str, threading.Thread] = {}
        
        # Register message handler
        self.device_management.register_message_handler('wifi', self._send_to_device)
        
        self.logger.info("WiFi transport initialized",
                        host=config.wifi.server_host,
                        port=config.wifi.server_port)
    
    def start(self) -> bool:
        """Start the WiFi transport"""
        if self.running:
            return True
        
        try:
            # Create and bind server socket
            self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.server_socket.bind((self.config.wifi.server_host, self.config.wifi.server_port))
            self.server_socket.listen(self.config.wifi.max_connections)
            
            self.running = True
            
            # Start accept thread
            self.accept_thread = threading.Thread(
                target=self._accept_connections,
                daemon=True,
                name="WiFiAccept"
            )
            self.accept_thread.start()
            
            self.logger.info("WiFi transport started",
                           listening=f"{self.config.wifi.server_host}:{self.config.wifi.server_port}")
            return True
            
        except Exception as e:
            self.logger.error(f"Failed to start WiFi transport: {e}", exc_info=True)
            self.stop()
            return False
    
    def stop(self):
        """Stop the WiFi transport"""
        if not self.running:
            return
        
        self.logger.info("Stopping WiFi transport...")
        self.running = False
        
        # Close server socket
        if self.server_socket:
            try:
                self.server_socket.close()
            except:
                pass
        
        # Stop all client handlers
        for handler in self.client_handlers.values():
            handler.stop()
        
        # Wait for accept thread
        if self.accept_thread and self.accept_thread.is_alive():
            self.accept_thread.join(timeout=5)
        
        # Wait for client threads
        for thread in self.client_threads.values():
            if thread.is_alive():
                thread.join(timeout=2)
        
        self.client_handlers.clear()
        self.client_threads.clear()
        
        self.logger.info("WiFi transport stopped")
    
    def is_running(self) -> bool:
        """Check if transport is running"""
        return self.running
    
    def _accept_connections(self):
        """Accept incoming client connections"""
        self.logger.debug("WiFi accept thread started")
        
        while self.running:
            try:
                client_socket, client_address = self.server_socket.accept()
                
                if not self.running:
                    client_socket.close()
                    break
                
                client_key = f"{client_address[0]}:{client_address[1]}"
                
                self.logger.info(f"New WiFi client connection", address=str(client_address))
                
                # Create client handler
                handler = WiFiClientHandler(
                    client_socket=client_socket,
                    client_address=client_address,
                    config=self.config,
                    routing_service=self.routing_service,
                    device_management=self.device_management,
                    statistics_service=self.statistics_service
                )
                
                # Start client thread
                client_thread = threading.Thread(
                    target=handler.start,
                    daemon=True,
                    name=f"WiFiClient-{client_address[0]}"
                )
                
                self.client_handlers[client_key] = handler
                self.client_threads[client_key] = client_thread
                client_thread.start()
                
                # Clean up finished threads
                self._cleanup_finished_threads()
                
            except Exception as e:
                if self.running:
                    self.logger.error(f"Error accepting connection: {e}")
                    time.sleep(1)
    
    def _cleanup_finished_threads(self):
        """Clean up finished client threads"""
        finished_keys = []
        for key, thread in self.client_threads.items():
            if not thread.is_alive():
                finished_keys.append(key)
        
        for key in finished_keys:
            self.client_handlers.pop(key, None)
            self.client_threads.pop(key, None)
    
    def _send_to_device(self, device_id: str, packet: ProtocolPacket) -> bool:
        """Send a packet to a specific device"""
        # Find client handler for device
        for handler in self.client_handlers.values():
            if handler.device_id == device_id:
                return handler.send_message(packet)
        
        self.logger.warning(f"No WiFi connection found for device {device_id}")
        return False
    
    def get_connected_devices(self) -> Dict[str, Any]:
        """Get list of connected WiFi devices"""
        devices = {}
        for handler in self.client_handlers.values():
            if handler.device_id and handler.assigned_address is not None:
                devices[handler.device_id] = {
                    'address': f"0x{handler.assigned_address:04X}",
                    'client_ip': handler.client_address[0],
                    'client_port': handler.client_address[1],
                    'interface': 'wifi'
                }
        return devices


def get_wlan0_ip() -> str:
    """Get the IP address of wlan0 interface"""
    try:
        result = subprocess.run(['ip', 'addr', 'show', 'wlan0'],
                              capture_output=True, text=True)

        if result.returncode == 0:
            for line in result.stdout.split('\n'):
                if 'inet ' in line and 'scope global' in line:
                    ip = line.strip().split()[1].split('/')[0]
                    return ip
    except Exception as e:
        pass

    # Fallback to default AP address
    return "192.168.50.1"

File: router/src/transports/wifi_transport.cpp
#include "transports/wifi_transport.hpp"
#include "transports/wifi_client_handler.hpp"
#include "core/config.hpp"
#include "core/logger.hpp"
#include "services/routing_service.hpp"
#include "services/device_management_service.hpp"
#include "services/statistics_service.hpp"
#include "config/mita_config.h"
#include "transport/transport_constants.h"

#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/ip.h>
#include <sys/select.h>
#include <fcntl.h>
#include <errno.h>
#include <cstring>
#include <vector>
#include <ifaddrs.h>
#include <net/if.h>

namespace mita
{
    namespace transports
    {

        // WiFiTransport implementation
        WiFiTransport::WiFiTransport(const core::RouterConfig &config,
                                     services::RoutingService &routing_service,
                                     services::DeviceManagementService &device_management,
                                     services::StatisticsService &statistics_service,
                                     std::shared_ptr<services::PacketMonitorService> packet_monitor)
            : BaseTransport(config, routing_service, device_management, statistics_service),
              raw_socket_(-1),
              logger_(core::get_logger("WiFiTransport")),
              packet_monitor_(packet_monitor)
        {
            // Get WiFi AP IP from config
            wifi_ap_ip_ = config_.wifi.server_host;
            
            logger_->info("WiFi transport initialized",
                          core::LogContext()
                              .add("protocol", MITA_IP_PROTOCOL)
                              .add("ap_ip", wifi_ap_ip_));
        }

        WiFiTransport::~WiFiTransport()
        {
            stop();
        }

        bool WiFiTransport::start()
        {
            if (running_.exchange(true))
            {
                return true; // Already running
            }

            try
            {
                // Create raw socket for custom IP protocol
                raw_socket_ = socket(AF_INET, SOCK_RAW, MITA_IP_PROTOCOL);
                if (raw_socket_ < 0)
                {
                    logger_->error("Failed to create raw socket",
                                   core::LogContext().add("error", strerror(errno)).add("protocol", MITA_IP_PROTOCOL));
                    return false;
                }

                // Set socket to non-blocking
                int flags = fcntl(raw_socket_, F_GETFL, 0);
                fcntl(raw_socket_, F_SETFL, flags | O_NONBLOCK);

                // Enable IP header inclusion for sending
                int one = 1;
                if (setsockopt(raw_socket_, IPPROTO_IP, IP_HDRINCL, &one, sizeof(one)) < 0)
                {
                    logger_->warning("Failed to set IP_HDRINCL",
                                     core::LogContext().add("error", strerror(errno)));
                }

                // Get local IP address
                struct ifaddrs *ifaddr, *ifa;
                if (getifaddrs(&ifaddr) == -1)
                {
                    logger_->warning("Failed to get network interfaces",
                                     core::LogContext().add("error", strerror(errno)));
                    local_ip_ = "0.0.0.0"; // Fallback
                }
                else
                {
                    // Find first non-loopback IPv4 address
                    for (ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next)
                    {
                        if (ifa->ifa_addr == nullptr)
                            continue;

                        if (ifa->ifa_addr->sa_family == AF_INET)
                        {
                            struct sockaddr_in *addr = (struct sockaddr_in *)ifa->ifa_addr;
                            char ip_str[INET_ADDRSTRLEN];
                            inet_ntop(AF_INET, &addr->sin_addr, ip_str, INET_ADDRSTRLEN);

                            // Skip loopback
                            if (strcmp(ip_str, "127.0.0.1") != 0)
                            {
                                local_ip_ = ip_str;
                                break;
                            }
                        }
                    }
                    freeifaddrs(ifaddr);

                    if (local_ip_.empty())
                    {
                        local_ip_ = "0.0.0.0"; // Fallback
                    }
                }

                logger_->info("Using local IP for raw packets",
                              core::LogContext().add("local_ip", local_ip_));

                // Start receive thread
                receive_thread_ = std::make_unique<std::thread>(&WiFiTransport::receive_packets, this);

                // Register message handler
                device_management_.register_message_handler("wifi", [this](const std::string &device_id, const protocol::ProtocolPacket &packet)
                                                            { send_packet(device_id, packet); });

                logger_->info("WiFi transport started",
                              core::LogContext().add("protocol", MITA_IP_PROTOCOL));
                return true;
            }
            catch (const std::exception &e)
            {
                logger_->error("Exception starting WiFi transport",
                               core::LogContext().add("error", e.what()));
                running_ = false;
                return false;
            }
        }

        void WiFiTransport::stop()
        {
            if (!running_.exchange(false))
            {
                return; // Already stopped
            }

            // Close raw socket to stop receiving
            if (raw_socket_ >= 0)
            {
                close(raw_socket_);
                raw_socket_ = -1;
            }

            // Wait for receive thread
            if (receive_thread_ && receive_thread_->joinable())
            {
                receive_thread_->join();
            }

            // Stop all client handlers
            {
                std::lock_guard<std::recursive_mutex> lock(clients_mutex_);
                for (auto &[device_id, handler] : client_handlers_)
                {
                    handler->stop();
                }
                client_handlers_.clear();
                ip_to_device_.clear();
            }

            logger_->info("WiFi transport stopped");
        }

        bool WiFiTransport::send_packet(const std::string &device_id, const protocol::ProtocolPacket &packet)
        {
            logger_->debug("WiFiTransport::send_packet called",
                           core::LogContext()
                               .add("device_id", device_id)
                               .add("msg_type", static_cast<int>(packet.get_message_type()))
                               .add("dest", packet.get_dest_addr())
                               .add("src", packet.get_source_addr()));

            std::lock_guard<std::recursive_mutex> lock(clients_mutex_);

            auto *handler = find_client_handler(device_id);
            if (!handler)
            {
                logger_->warning("Cannot send packet - device not found",
                                 core::LogContext()
                                     .add("device_id", device_id)
                                     .add("available_handlers", client_handlers_.size()));

                // Log all available device IDs
                for (const auto &[id, h] : client_handlers_)
                {
                    logger_->debug("Available client handler",
                                   core::LogContext().add("device_id", id));
                }
                return false;
            }

            logger_->debug("WiFiTransport::send_packet - found handler, sending",
                           core::LogContext().add("device_id", device_id));

            bool result = handler->send_packet(packet);
            logger_->debug("WiFiTransport::send_packet result",
                           core::LogContext()
                               .add("device_id", device_id)
                               .add("result", result));
            return result;
        }

        int WiFiTransport::broadcast_packet(const protocol::ProtocolPacket &packet)
        {
            std::lock_guard<std::recursive_mutex> lock(clients_mutex_);

            int sent_count = 0;
            for (auto &[device_id, handler] : client_handlers_)
            {
                if (handler->is_authenticated() && handler->send_packet(packet))
                {
                    sent_count++;
                }
            }

            return sent_count;
        }

        std::string WiFiTransport::get_connection_info() const
        {
            return "WiFi Raw IP - Protocol " + std::to_string(MITA_IP_PROTOCOL);
        }

        void WiFiTransport::receive_packets()
        {
            logger_->info("WiFi transport receiving packets");

            auto last_cleanup = std::chrono::steady_clock::now();
            const auto cleanup_interval = std::chrono::seconds(5);

            const size_t max_packet_size = 65535;
            uint8_t *buffer = new uint8_t[max_packet_size];

            while (running_)
            {
                auto now = std::chrono::steady_clock::now();
                if (now - last_cleanup >= cleanup_interval)
                {
                    cleanup_disconnected_clients();
                    last_cleanup = now;
                }

                struct sockaddr_in src_addr;
                socklen_t addr_len = sizeof(src_addr);

                ssize_t received = recvfrom(raw_socket_, buffer, max_packet_size, 0,
                                            (struct sockaddr *)&src_addr, &addr_len);

                if (received > 0)
                {
                    // Extract IP header
                    struct ip *ip_hdr = (struct ip *)buffer;
                    uint8_t ihl = ip_hdr->ip_hl * 4; // Header length in bytes

                    // Verify it's our protocol
                    if (ip_hdr->ip_p == MITA_IP_PROTOCOL && received > ihl)
                    {
                        // Extract source IP
                        char source_ip_str[INET_ADDRSTRLEN];
                        inet_ntop(AF_INET, &(ip_hdr->ip_src), source_ip_str, INET_ADDRSTRLEN);
                        std::string source_ip(source_ip_str);

                        // Extract payload (skip IP header)
                        const uint8_t *payload = buffer + ihl;
                        size_t payload_len = received - ihl;

                        handle_packet_from_ip(source_ip, payload, payload_len);
                    }
                }
                else if (received < 0 && errno != EAGAIN && errno != EWOULDBLOCK)
                {
                    if (running_)
                    {
                        logger_->error("recvfrom error",
                                       core::LogContext().add("error", strerror(errno)));
                    }
                }

                // Small delay to avoid busy waiting
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }

            delete[] buffer;
            logger_->debug("WiFi transport stopped receiving");
        }

        void WiFiTransport::handle_packet_from_ip(const std::string &source_ip, const uint8_t *data, size_t length)
        {
            // Deserialize the packet
            auto packet_ptr = protocol::ProtocolPacket::from_bytes(data, length);
            if (!packet_ptr)
            {
                logger_->warning("Failed to deserialize packet",
                                 core::LogContext().add("source_ip", source_ip));
                return;
            }

            protocol::ProtocolPacket &packet = *packet_ptr;

            // Check if this is a HELLO packet (new connection)
            if (packet.get_message_type() == MessageType::HELLO)
            {
                std::string router_id, device_id;
                std::vector<uint8_t> nonce1;
                if (!protocol::utils::parse_hello_packet(packet, router_id, device_id, nonce1))
                {
                    logger_->warning("Failed to parse HELLO packet",
                                     core::LogContext().add("source_ip", source_ip));
                    return;
                }
                std::lock_guard<std::recursive_mutex> lock(clients_mutex_);

                // Check if we already have a handler for this device
                auto it = client_handlers_.find(device_id);
                if (it == client_handlers_.end())
                {
                    // New device - check connection limit
                    if (client_handlers_.size() >= static_cast<size_t>(config_.wifi.max_connections))
                    {
                        logger_->warning("Connection limit reached, rejecting client",
                                         core::LogContext().add("device_id", device_id).add("source_ip", source_ip));
                        return;
                    }

                    // Create new handler
                    try
                    {
                        auto handler = std::make_unique<WiFiClientHandler>(
                            source_ip, config_,
                            routing_service_, device_management_, statistics_service_, packet_monitor_);

                        // Set send callback
                        handler->set_send_callback([this](const std::string &ip, const uint8_t *data, size_t len)
                                                   { return send_raw_packet(ip, data, len); });

                        client_handlers_[device_id] = std::move(handler);
                        ip_to_device_[source_ip] = device_id;

                        client_handlers_[device_id]->start(packet);

                        logger_->info("New WiFi client connected",
                                      core::LogContext().add("device_id", device_id).add("source_ip", source_ip));
                    }
                    catch (const std::exception &e)
                    {
                        logger_->error("Exception creating client handler",
                                       core::LogContext().add("device_id", device_id).add("error", e.what()));
                        statistics_service_.record_transport_error("wifi");
                    }
                }
                else
                {
                    // Existing device - update IP if changed
                    ip_to_device_[source_ip] = device_id;
                    it->second->handle_packet(packet);
                }
            }
            else
            {
                // Regular packet - find handler by IP
                std::lock_guard<std::recursive_mutex> lock(clients_mutex_);

                auto handler = find_client_by_ip(source_ip);
                if (handler)
                {
                    handler->handle_packet(packet);
                }
                else
                {
                    logger_->warning("Received packet from unknown source",
                                     core::LogContext().add("source_ip", source_ip));
                }
            }
        }

        bool WiFiTransport::send_raw_packet(const std::string &dest_ip, const uint8_t *data, size_t length)
        {
            if (raw_socket_ < 0)
            {
                return false;
            }

            // Calculate total packet size (IP header + data)
            const size_t ip_header_len = 20;
            size_t total_len = ip_header_len + length;
            uint8_t *packet = new uint8_t[total_len];

            // Build IP header
            memset(packet, 0, ip_header_len);

            // Version (4) and IHL (5 = 20 bytes)
            packet[0] = 0x45;

            // Type of Service
            packet[1] = 0;

            // Total Length
            packet[2] = (total_len >> 8) & 0xFF;
            packet[3] = total_len & 0xFF;

            // Identification
            static uint16_t ip_id = 0;
            ip_id++;
            packet[4] = (ip_id >> 8) & 0xFF;
            packet[5] = ip_id & 0xFF;

            // Flags and Fragment Offset
            packet[6] = 0x40; // Don't fragment
            packet[7] = 0;

            // TTL
            packet[8] = 64;

            // Protocol
            packet[9] = MITA_IP_PROTOCOL;

            // Header Checksum (will be calculated)
            packet[10] = 0;
            packet[11] = 0;

            // Determine source IP based on destination subnet
            // If destination is on WiFi AP subnet (192.168.50.x), use WiFi AP IP as source
            std::string src_ip;
            if (dest_ip.substr(0, 11) == "192.168.50.")
            {
                src_ip = wifi_ap_ip_;  // Use WiFi AP interface IP
            }
            else
            {
                src_ip = local_ip_;  // Use main network interface IP
            }

            // Source IP
            struct in_addr src_addr;
            inet_pton(AF_INET, src_ip.c_str(), &src_addr);
            memcpy(packet + 12, &src_addr, 4);

            // Destination IP
            struct in_addr dest_addr;
            inet_pton(AF_INET, dest_ip.c_str(), &dest_addr);
            memcpy(packet + 16, &dest_addr, 4);

            // Calculate IP header checksum
            uint32_t sum = 0;
            for (size_t i = 0; i < ip_header_len; i += 2)
            {
                uint16_t word = (packet[i] << 8) | packet[i + 1];
                sum += word;
            }
            while (sum >> 16)
            {
                sum = (sum & 0xFFFF) + (sum >> 16);
            }
            uint16_t checksum = ~sum;
            packet[10] = (checksum >> 8) & 0xFF;
            packet[11] = checksum & 0xFF;

            // Copy payload
            memcpy(packet + ip_header_len, data, length);

            // Send packet
            struct sockaddr_in dest_sockaddr;
            memset(&dest_sockaddr, 0, sizeof(dest_sockaddr));
            dest_sockaddr.sin_family = AF_INET;
            dest_sockaddr.sin_addr = dest_addr;

            ssize_t sent = sendto(raw_socket_, packet, total_len, 0,
                                  (struct sockaddr *)&dest_sockaddr, sizeof(dest_sockaddr));

            delete[] packet;

            if (sent < 0)
            {
                logger_->error("sendto failed",
                               core::LogContext().add("error", strerror(errno)).add("dest_ip", dest_ip));
                return false;
            }

            logger_->debug("Raw packet sent successfully",
                           core::LogContext()
                               .add("dest_ip", dest_ip)
                               .add("src_ip", src_ip)
                               .add("payload_size", length)
                               .add("total_size", total_len)
                               .add("sent_bytes", sent));

            return sent == (ssize_t)total_len;
        }

        void WiFiTransport::cleanup_disconnected_clients()
        {
            {
                std::lock_guard<std::recursive_mutex> lock(clients_mutex_);
                for (auto& pair : client_handlers_)
                {
                    if (pair.second)
                    {
                        pair.second->check_heartbeat_timeout();
                    }
                }
            }

            std::vector<std::unique_ptr<WiFiClientHandler>> handlers_to_cleanup;

            {
                std::lock_guard<std::recursive_mutex> lock(clients_mutex_);

                int cleaned = 0;
                auto it = client_handlers_.begin();
                while (it != client_handlers_.end())
                {
                    if (it->second && it->second->check_for_disconnected())
                    {
                        std::string device_id = it->first;
                        std::string client_ip = it->second->get_client_ip();

                        logger_->info("Removing disconnected client",
                                      core::LogContext().add("device_id", device_id).add("ip", client_ip));

                        if (!device_id.empty())
                        {
                            device_management_.remove_device(device_id);
                        }

                        // Remove IP mapping
                        ip_to_device_.erase(client_ip);

                        handlers_to_cleanup.push_back(std::move(it->second));
                        it = client_handlers_.erase(it);
                        cleaned++;
                    }
                    else
                    {
                        ++it;
                    }
                }

                if (cleaned > 0)
                {
                    logger_->info("Cleaned up disconnected clients",
                                  core::LogContext().add("cleaned_count", cleaned).add("remaining", client_handlers_.size()));
                }
            }

            handlers_to_cleanup.clear();
        }

        std::vector<WiFiClientHandler*> WiFiTransport::get_all_client_handlers() const
        {
            std::lock_guard<std::recursive_mutex> lock(clients_mutex_);
            std::vector<WiFiClientHandler*> handlers;

            for (const auto& device : client_handlers_)
            {
                if (device.second)
                {
                    handlers.push_back(device.second.get());
                }
            }

            return handlers;
        }

        WiFiClientHandler *WiFiTransport::find_client_handler(const std::string &device_id)
        {
            auto it = client_handlers_.find(device_id);
            if (it != client_handlers_.end())
            {
                return it->second.get();
            }
            return nullptr;
        }

        WiFiClientHandler *WiFiTransport::find_client_by_ip(const std::string &ip_address)
        {
            auto it = ip_to_device_.find(ip_address);
            if (it != ip_to_device_.end())
            {
                return find_client_handler(it->second);
            }
            return nullptr;
        }

    } // namespace transports
} // namespace mita